sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,691 @@
+    {
+      resp_data.content_type = az_mqtt5_property_get_string(content_type);
+    }
+  }
+
+  return resp_data;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_rpc_client_hfsm* this_policy = (az_mqtt5_rpc_client_hfsm*)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_client_hfsm"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+    {
+      _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
+          this_policy->_internal.connection,
+          (az_event){ .type = AZ_HFSM_EVENT_ERROR, .data = event.data }));
+      // TODO: uncomment when this won't fail every time
+      // if (az_result_failed(az_event_policy_send_inbound_event(me, event)))
+      // {
+      //   az_platform_critical_error();
+      // }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_EXIT:
+    {
+      if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_EXIT))
+      {
+        _az_LOG_WRITE(AZ_HFSM_EVENT_EXIT, AZ_SPAN_FROM_STR("az_mqtt5_rpc_client_hfsm: PANIC!"));
+      }
+
+      az_platform_critical_error();
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+    case AZ_MQTT5_EVENT_UNSUBACK_RSP:
+      break;
+
+    case AZ_EVENT_RPC_CLIENT_UNSUB_REQ:
+    {
+      // Send unsubscribe
+      az_mqtt5_unsub_data unsubscription_data
+          = { .topic_filter = this_policy->_internal.rpc_client->response_topic };
+
+      _az_RETURN_IF_FAILED(az_event_policy_send_outbound_event(
+          (az_event_policy*)this_policy,
+          (az_event){ .type = AZ_MQTT5_EVENT_UNSUB_REQ, .data = &unsubscription_data }));
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUB_RECV_IND:
+    {
+      az_mqtt5_recv_data* recv_data = (az_mqtt5_recv_data*)event.data;
+
+      // Ensure pub is of the right topic
+      if (az_span_is_content_equal(
+              this_policy->_internal.rpc_client->response_topic, recv_data->topic))
+      {
+        az_mqtt5_property_binarydata correlation_data;
+        az_mqtt5_property_stringpair status;
+        az_mqtt5_property_stringpair error_message;
+        az_mqtt5_property_string content_type;
+
+        az_mqtt5_rpc_client_rsp_event_data resp_data
+            = _parse_response(recv_data, &correlation_data, &status, &error_message, &content_type);
+
+        // send to application to handle
+        // if ((az_event_policy*)this_policy->inbound_policy != NULL)
+        // {
+        // az_event_policy_send_inbound_event((az_event_policy*)this_policy, (az_event){.type =
+        // AZ_EVENT_RPC_SERVER_EXECUTE_COMMAND_REQ, .data = data});
+        // }
+        _az_RETURN_IF_FAILED(_az_mqtt5_connection_api_callback(
```

Comment:
If this returns, none of the free statements are executed.
I think we need to redesign around an inner state that has the sole purpose of performing allocations and de-allocations.

Reply:
fixed by not returning here if this fails